api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+		// The SDK states that you can specify up to 10 services in one DescribeServices operation:
+		// https://github.com/aws/aws-sdk-go/blob/ee1f179877b2daf2aaabf71fa900773bf8842253/service/ecs/api.go#L5420
+		// (aws-sdk-go version 1.12.19, as stated in layer0/Gopkg.toml)
+		for i := 0; i < len(serviceARNs); i += 10 {
+			end := i + 10
+
+			if end > len(serviceARNs) {
+				end = len(serviceARNs)
+			}
+
+			describeServicesInput := &ecs.DescribeServicesInput{}
+			describeServicesInput.SetCluster(clusterName)
+			describeServicesInput.SetServices(serviceARNs[i:end])
+
+			output, err := e.Client.ECS.DescribeServices(describeServicesInput)
+			if err != nil {
+				return nil, err
+			}
+
+			services = append(services, output.Services...)
+		}
+	}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int64{}
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
+			runningCount := aws.Int64Value(d.RunningCount)
+			pendingCount := aws.Int64Value(d.PendingCount)
+			if numPending := desiredCount - (runningCount + pendingCount); numPending > 0 {
+				deployIDCopies[aws.StringValue(d.Id)] = numPending
+			}
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys
+		for deployID := range deployIDCopies {
+			c, err := e.getContainerResourceFromDeploy(deployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, c...)
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers_TasksInECS(clusterName string) ([]ResourceConsumer, error) {
+	var (
+		taskARNs          []string
+		resourceConsumers []ResourceConsumer
+	)
+
+	fn := func(output *ecs.ListTasksOutput, lastPage bool) bool {
+
+		for _, taskARN := range output.TaskArns {
+			taskARNs = append(taskARNs, aws.StringValue(taskARN))
+		}
+
+		return !lastPage
+	}
+
+	startedBy := e.Config.Instance()
+	for _, status := range []string{ecs.DesiredStatusRunning, ecs.DesiredStatusStopped} {
+		input := &ecs.ListTasksInput{}
+		input.SetCluster(clusterName)
+		input.SetDesiredStatus(status)
+		input.SetStartedBy(startedBy)
+		if err := e.Client.ECS.ListTasksPages(input, fn); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, taskARN := range taskARNs {
+		task, err := e.TaskProvider.Read(taskARN)
+		if err != nil {
+			return nil, err
+		}
+
+		taskResourceConsumers, err := e.getContainerResourceFromDeploy(task.DeployID)
+		if err != nil {
+			return nil, err
+		}
+
+		resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers_TasksInJobs(clusterName string) ([]ResourceConsumer, error) {
+	var resourceConsumers []ResourceConsumer
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateTaskJob {
+			// TODO: maybe remove pending check here
```

Comment:
why (maybe) remove pending check?

Reply:
This was recommend by me. The more I think about it, the more I'm convinced we should *only* check if the job status is `InProgress`. It won't make much of a difference for our use case, but if you think about extremely high scale, it's more efficient and cost-effective to only scale for jobs that are `InProgress`. Jobs could be stuck in the `Pending` state for hours waiting for a worker to pick them up.

Reply:
✅